src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingSourceGenerator.Parser.cs

```diff
@@ -922,29 +964,59 @@ public static bool ContainsGenericParameters(INamedTypeSymbol type)
                 return false;
             }
 
-            private static bool HasPublicParameterlessCtor(INamedTypeSymbol type)
+            private static bool HasPublicParameterLessCtor(INamedTypeSymbol type) =>
+                CannotBeConstructed(type)
+                ? false
+                : type.InstanceConstructors.SingleOrDefault(ctor => ctor.DeclaredAccessibility is Accessibility.Public && ctor.Parameters.Length is 0)
+                    is not null;
+
+            private static IMethodSymbol? GetConstructor(INamedTypeSymbol type)
             {
-                if (type.IsAbstract || type.TypeKind == TypeKind.Interface)
+                if (type.Name is "ReadonlyRecordStructTypeOptions")
                 {
-                    return false;
                 }
 
-                if (type is not INamedTypeSymbol namedType)
+                if (CannotBeConstructed(type))
                 {
-                    return false;
+                    return null;
                 }
 
-                foreach (IMethodSymbol ctor in namedType.InstanceConstructors)
+                IMethodSymbol? publicParameterlessCtor = null;
+                IMethodSymbol? publicParameterizedCtor = null;
+                bool hasMultiplePublicParameterizedCtors = false;
+
+                foreach (IMethodSymbol ctor in type.InstanceConstructors)
                 {
-                    if (ctor.DeclaredAccessibility == Accessibility.Public && ctor.Parameters.Length == 0)
+                    if (ctor.DeclaredAccessibility is not Accessibility.Public)
```

Comment:
Do we want to support internal constructors? Instead of evaluating the declared accessibility you could try using the [`Compilation.IsSymbolAccessibleWithin`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.compilation.issymbolaccessiblewithin?view=roslyn-dotnet-4.3.0) which lets you know if a particular symbol is accessible from the type that is being generated.